packages/next/src/server/app-render/collect-segment-data.tsx

```diff
@@ -0,0 +1,309 @@
+  clientModules: ManifestNode,
+  serverConsumerManifest: any,
+  segmentPath: Array<[string, Segment]>,
+  segmentPathStr: string,
+  segmentBufferMap: Map<string, Buffer>,
+  collectedTasks: Array<Promise<void>>
+): void {
+  const children = route[1]
+  for (const parallelRouteKey in children) {
+    const childRoute = children[parallelRouteKey]
+    const childSegment = childRoute[0]
+    const childSegmentPath = segmentPath.concat([
+      [parallelRouteKey, childSegment],
+    ])
+    const childSegmentPathStr =
+      segmentPathStr +
+      '/' +
+      encodeChildSegmentAsFilesystemSafePathname(parallelRouteKey, childSegment)
+    collectSegmentDataImpl(
+      childRoute,
+      fullPageDataBuffer,
+      clientModules,
+      serverConsumerManifest,
+      childSegmentPath,
+      childSegmentPathStr,
+      segmentBufferMap,
+      collectedTasks
+    )
+  }
+
+  // Spawn a task to render the segment data to a stream.
+  collectedTasks.push(
+    renderSegmentDataToStream(
+      fullPageDataBuffer,
+      clientModules,
+      serverConsumerManifest,
+      segmentPath,
+      segmentPathStr,
+      segmentBufferMap
+    )
+  )
+}
+
+async function renderSegmentDataToStream(
+  fullPageDataBuffer: Buffer,
+  clientModules: ManifestNode,
+  serverConsumerManifest: any,
+  segmentPath: Array<[string, Segment]>,
+  segmentPathStr: string,
+  segmentBufferMap: Map<string, Buffer>
+) {
+  // Create a new Flight response that contains data only for this segment.
+  try {
+    // Since all we're doing is decoding and re-encoding a cached prerender, if
+    // it takes longer than a microtask, it must because of hanging promises
+    // caused by dynamic data. Abort the stream at the end of the current task.
+    const abortController = new AbortController()
+    waitAtLeastOneReactRenderTask().then(() => abortController.abort())
+
+    const segmentStream = renderToReadableStream(
+      <PickSegment
+        fullPageDataBuffer={fullPageDataBuffer}
+        serverConsumerManifest={serverConsumerManifest}
+        segmentPath={segmentPath}
+      />,
+      clientModules,
+      {
+        signal: abortController.signal,
+        onError() {
+          // Ignore any errors. These would have already been reported when
+          // we created the full page data.
+        },
+      }
+    )
+    const segmentBuffer = await streamToBuffer(segmentStream)
+    // Add the buffer to the result map.
+    if (segmentPathStr === '') {
+      segmentBufferMap.set('/', segmentBuffer)
+    } else {
+      segmentBufferMap.set(segmentPathStr, segmentBuffer)
+    }
+  } catch {
+    // If there are any errors, then we skip the segment. The effect is that
+    // a prefetch for this segment will 404.
+  }
+}
+
+async function PickSegment({
+  fullPageDataBuffer,
+  serverConsumerManifest,
+  segmentPath,
+}: {
+  fullPageDataBuffer: Buffer
+  serverConsumerManifest: any
+  segmentPath: Array<[string, Segment]>
+}): Promise<any> {
```

Comment:
The type system doesn't allow you to return arbitrary data from a React component. It expects a ReactNode. It works in the runtime, though; since this is a Server Component, in the resulting stream the element is transparent.

The reason I have to wrap it in a component is so that when we decode the original stream, the side effects are transferred to the new stream.

But to make it clearer I'll move the escape hatch to where the element is created, and add a comment.